### DISCUSSION TITLE
Removing usage of IAM AWS Managed Policies in L2+ constructs

### DISCUSSION THREAD

**Discussion by dontirun** [aws/aws-cdk#22123] [General]
I think it would be ultimately beneficial to replace AWS Managed Policies in L2+ constructs with Customer Managed Policies/Inline Policies and would like to hear other people's thoughts on the matter

## Background

I've had many cases where I've been asked to remove AWS Managed Policies from L2+ constructs. While this can be done through Aspects and Overrides this is both tedious and not friendly.

## Rationale

There are 2 main problems using AWS managed policies 

1. When AWS updates the policies, the permissions granted to services using those policies change without customer knowledge and outside of deployment
2. Generally [aren't designed for least privileged access](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_managed-vs-inline.html#aws-or-customer-managed)

## Pros
1.  User control over permission changes
2.  Better scoped privileges

## Cons
1. Increased template size

**Comment by niebloomj** [comment]
Cdknag is always complaining about this. I definitely agree with the approach here. We need to be building with least privilege in mind by default! 🚀🥳🚀

**Comment by maafk** [comment]
In the [AWS Solutions nag pack](https://github.com/cdklabs/cdk-nag/blob/main/RULES.md#aws-solutions), `AwsSolutions-IAM4` errors due to AWS Managed Policies being used

> An AWS managed policy is a standalone policy that is created and administered by AWS. Currently, many AWS managed policies do not restrict resource scope. Replace AWS managed policies with system specific (customer) managed policies. This is a granular rule that returns individual findings that can be suppressed with appliesTo. The findings are in the format Policy::<policy> for AWS managed policies. Example: appliesTo: ['Policy::arn:<AWS::Partition>:iam::aws:policy/foo'].

Since AWS managed policies are used in many L2/3 constructs ([example](https://github.com/aws/aws-cdk/blob/6840bc39a1d80ceb45da8f60c6e6b5c1243c31f8/packages/%40aws-cdk/aws-lambda/lib/function.ts#L682)), I end up adding dozens of suppressions for `AwsSolutions-IAM4`.

**Reply by niebloomj** [reply]
Same over here. I would love to stop having to code the suppressions into every single one of my projects

**Reply by dontirun** [reply]
I do want to make to note that cdk-nag isn't part of the CDK or maintained by the CDK team. I think not having to suppress the rule is a side benefit to fixing the actual problem

**Reply by niebloomj** [reply]
Right. It's definitely a side benefit. But the reason it's part of CDK nag is that it has a security smell. I definitely think that whatever our defaults are for using CDK should be at the highest security bar. Secure by default.
